### ISSUE TITLE
Duplicate collision detection when re-instancing Area2D nodes

### ISSUE THREAD

**Issue by skoovo** [godotengine/godot#34085]
**Godot version:**

3.2 BETA 2

**OS/device including version:**

Windows 10 Pro 64 bit 1909

**Issue description:**

I'm using a `KinematicBody2D` for a "player" node and `Area2D` nodes for "obstacles". The player moves with `move_and_collide` and when it collides with an obstacle, it is game over and the level is reset. At this point, the player node has it's `position` manually set to the spawn point, while the obstacles are re-instanced (In the context of a larger level, where the states of many objects can change, re-instancing the level is preferable to going back and resetting the states of each individual object).

The issue, is that the new `Area2D` obstacles are colliding with the player node that has already moved out of the way. I don't know the inner workings of the Godot engine or how its physics works, but it feels like something is out of sync at some point.

**Steps to reproduce:**

Collide a `KinematicBody2D` with an `Area2D`. If the `KinematicBody2D` is then manually moved away from the obstacle, and the obstacle left as is, the collision happens once as expected. If the `KinematicBody2D` is manually moved away from the obstacle and the obstacle is recreated, the collision fires on the new obstacle even though the player node has been moved.

**Minimal reproduction project:**

I have attached a minimal project to show this behaviour:
[CollisionTest.zip](https://github.com/godotengine/godot/files/3916999/CollisionTest.zip)

I posted this issue on the Discord, but we couldn't figure it out. It does seem strangely simple, so I apologise in advance if this turns out not to be a bug, but I'm just doing something really stupid :)

**Comment by Xrayez** [comment]
I guess you have several options:
1. Re-instance the player too instead of moving it.
2. Use deferring mechanics: `call_deferred`, `set_deferred`, `yield(get_tree(), "physics_frame")`:

```gdscript
	if Input.is_action_just_pressed("ui_down"):
		print("--- DOWN ---")
		reset_player()
		yield(get_tree(), "physics_frame")
		remove_obstacle()
		create_obstacle()
```
3. Use `NOTIFICATION_TRANSFORM_CHANGED` (actually doesn't fix the issue but may be of worth in other "respawn" cases)
```gdscript
extends KinematicBody2D

signal position_changed()

func _notification(what):
	match what:
		NOTIFICATION_TRANSFORM_CHANGED:
			emit_signal("position_changed")

# ...

yield(hero, "position_changed") # wait until hero is moved
```
For option (3) I personally hope for #6491 to be implemented. 🙂

Also see my concerns at #33610 regarding spawning behavior:

> There are a number of perfomance-critical aspects of spawning which can be solved, for instance how would you go about spawning collision bodies which don't collide with existing level geometry (as well as already-spawned bodies)?

I've personally have implemented some more complex spawning mechanics which required me to skip several frames until characters fit onto terrain colliders perfectly spaced without colliding each other, and I've used `Physics2DServer.body_test_motion` for that and the mentioned `yield`ing.

I'm not sure what are other ways to forcing the physics engine to update the real position of `KinematicBody2D` as it seems like updating character's transform doesn't update collision space immediately, but only on the next frame. There's `force_update_transform` but it doesn't do anything for physics.

**Comment by skoovo** [comment]
Thanks for taking the time to look into this. I tried `call_deferred` in various combinations and it didn't have an effect for me when used on `remove_obstacle` and `create_obstacle`. Your `yield` solution worked and only one signal was emitted. However, when I tried this in an actual project it failed.

`yield`ing a `physics_frame` only seems to work from within the `_physics_process` function. I'm using buttons as part of my UI, and when I reset the game via a `pressed` signal, yielding doesn't have any effect. Definitely seems to be some sort of sync issue with the physics.

I've updated my demo project to reflect my recent testing:
[CollisionTest.zip](https://github.com/godotengine/godot/files/3918411/CollisionTest.zip)

**Comment by Xrayez** [comment]
If you connect the signal with "Deferred" option, seems to resolve the issue as well, I'm not sure why:

![godot_deferred](https://user-images.githubusercontent.com/17108460/70085862-c139b700-1619-11ea-94c8-7dca67794ac9.png)

**Comment by skoovo** [comment]
Yes, deferring the signal in conjuction with `yield` in the signal's callback does work, but only sometimes. Timing seems to be involved too. If you click RESET a little while after the collision, it's more likely to work as expected. If you click it too soon after the collision, it emits from the new obstacle. It's almost like a lottery as to whether the new obstacle will also emit, which can't be right.

This intermittent behaviour has me really confused.

**Comment by KoBeWi** [comment]
> Yes, deferring the signal in conjuction with yield in the signal's callback does work, but only sometimes.

Not sure how much of a bug it is. It might be just the way Godot works. Correct me if I'm wrong, but "idle_frame" is independent of the physics. That's why when deferring the signal, it might be called before or after the physics frame and this determines if it works correctly or not. I guess a "clean" way to solve this would be adding a method that forces a position update for physics body.

For now, you can make a `respawned` bool variable inside player, which will be set to true when player resets and set to false on next player's `physics_update`. Then ignore the obstacle signal if `respawned` is set to true. It does sound like a hack, but when making games it's often impossible to avoid such workarounds without unusable "programming patterns".

**Comment by bojidar-bg** [comment]
Keep in mind that due to #34740, you might want to use `global_transform` / `global_position` after each `NOTIFICATION_TRANSFORM_CHANGED`.

**Comment by KoBeWi** [comment]
Still valid in 4d9b95f
